hibike/devices/BatteryBuzzer/voltage_sense.cpp

```diff
@@ -0,0 +1,212 @@
+    //   }
+    //   else
+    //   {
+    //     Serial.println("Single Calibration Mode");
+    //   }
+  
+    //   Serial.println("ABS: Cell1 : " + str_cell1 + " Cell2 : " + str_cell2 + " Cell3 : " + str_cell3);
+    //   Serial.println("REL: Cell1: " + str_cell1 + " Cell2 : " + sdv_cell2 + " Cell3 : " + sdv_cell3);
+    //   Serial.println("");
+    //   Serial.println("");
+
+
+    // }
+    last_print_time = millis();
+
+}
+
+void handle_calibration() //called very frequently by loop.
+{
+  if (digitalRead(calib_button)) //I pressed the button, start calibrating.
+    {
+    if(triple_calibration && calib[0] == 2.56 && calib[1] == 2.56 & calib[2] == 2.56)
+    {
+      //i'm in triple calibration mode, but the calibration array is exactly the same as the datasheet values... which means i haven't been calibrated.
+      float vref_new = calibrate();
+      write_triple_eeprom(calib[0],calib[1],calib[2]);
+      start_8_seg_sequence(2); 
+    }
+    else if(!triple_calibration && get_calibration() == -1.0) //i'm not in triple calibration mode, and i haven't been calibrated.
+    {   
+      float vref_new = calibrate();
+      write_single_eeprom(vref_new);
+      start_8_seg_sequence(2);
+    }
+    else //I have arleady been calibrated with something.  reset it.
+    {
+      //wait untill the button press goes away.
+      //This is to maintain commonality with how calibrate() works.
+      {
+        delay(1);
+      }
+      
+      if(prints)
+      {
+        Serial.println("Clear EEPROM");
+      }
+      //reset all my calibration values as well, so my calibration values that i'm using are in lockstep with the EEPROM.
+      
+      vref_guess = 2.56;
+      calib[0] = 2.56;
+      calib[1] = 2.56;
+      calib[2] = 2.56;
+
+
+      clear_eeprom();
+      start_8_seg_sequence(1);
+      
+
+    }
+  }
+}
+
+float calibrate() //calibrates the device.
+//assumes that a 5V line has been put to each battery cell.
+//it then calculates its best guess for vref.
+{
+    if(prints)
+    {
+      Serial.println("I Should calibrate now");
+    }
+
+    //wait untill the button press goes away.
+    //This is jank debouncing.
+    //Have to calibrate AFTER the button is pressed, not while.  Yes, it's wierd.  PCB issue.
+    while (digitalRead(calib_button))
+    {
+      delay(1);
+    }
+    delay(100);
+
+
+    //ok.  let's assume (for now) that every line has 5V put at it.
+    //I want to calculate the best-fit for the true vref.
+
+    //expected voltages, after the resistive dividers.
+    float expected_vc1 = 5 * 10.0 / (10.00 + 10.00);
+    float expected_vc2 = 5 * 10.0 / (30.00 + 10.00);
+    float expected_vc3 = 5 * 10.0 / (51.00 + 10.0);
+
+
+    int r_cell1 = analogRead(cell1);
+    int r_cell2 = analogRead(cell2);
+    int r_cell3 = analogRead(cell3);
+
+
+    //as per whiteboard math.  Is it 1023 or 1024?  I think it's 1023.
+    float vref1 = 1024 / (float(r_cell1)) * 10.0 / (10.0 + 10.00) * 5;
```

Comment:
More uses of magic numbers. Resistor values should have a single source of truth.